src/core/pointcloud/qgspointcloudlayerprofilegenerator.cpp

```diff
@@ -0,0 +1,554 @@
+  tolerance = pcGenerator->mTolerance;
+  pointSize = pcGenerator->mPointSize;
+  pointSizeUnit = pcGenerator->mPointSizeUnit;
+  pointSymbol = pcGenerator->mPointSymbol;
+  pointColor = pcGenerator->mPointColor;
+  opacityByDistanceEffect = pcGenerator->mOpacityByDistanceEffect;
+}
+
+//
+// QgsPointCloudLayerProfileGenerator
+//
+
+QgsPointCloudLayerProfileGenerator::QgsPointCloudLayerProfileGenerator( QgsPointCloudLayer *layer, const QgsProfileRequest &request )
+  : mLayer( layer )
+  , mRenderer( mLayer->renderer() ? mLayer->renderer()->clone() : nullptr )
+  , mMaximumScreenError( qgis::down_cast< QgsPointCloudLayerElevationProperties* >( layer->elevationProperties() )->maximumScreenError() )
+  , mMaximumScreenErrorUnit( qgis::down_cast< QgsPointCloudLayerElevationProperties* >( layer->elevationProperties() )->maximumScreenErrorUnit() )
+  , mPointSize( qgis::down_cast< QgsPointCloudLayerElevationProperties* >( layer->elevationProperties() )->pointSize() )
+  , mPointSizeUnit( qgis::down_cast< QgsPointCloudLayerElevationProperties* >( layer->elevationProperties() )->pointSizeUnit() )
+  , mPointSymbol( qgis::down_cast< QgsPointCloudLayerElevationProperties* >( layer->elevationProperties() )->pointSymbol() )
+  , mPointColor( qgis::down_cast< QgsPointCloudLayerElevationProperties* >( layer->elevationProperties() )->pointColor() )
+  , mOpacityByDistanceEffect( qgis::down_cast< QgsPointCloudLayerElevationProperties* >( layer->elevationProperties() )->applyOpacityByDistanceEffect() )
+  , mId( layer->id() )
+  , mFeedback( std::make_unique< QgsFeedback >() )
+  , mProfileCurve( request.profileCurve() ? request.profileCurve()->clone() : nullptr )
+  , mTolerance( request.tolerance() )
+  , mSourceCrs( layer->crs() )
+  , mTargetCrs( request.crs() )
+  , mTransformContext( request.transformContext() )
+  , mZOffset( layer->elevationProperties()->zOffset() )
+  , mZScale( layer->elevationProperties()->zScale() )
+  , mStepDistance( request.stepDistance() )
+{
+  if ( mLayer->dataProvider()->index() )
+  {
+    mScale = mLayer->dataProvider()->index()->scale();
+    mOffset = mLayer->dataProvider()->index()->offset();
+  }
+}
+
+QString QgsPointCloudLayerProfileGenerator::sourceId() const
+{
+  return mId;
+}
+
+Qgis::ProfileGeneratorFlags QgsPointCloudLayerProfileGenerator::flags() const
+{
+  return Qgis::ProfileGeneratorFlag::RespectsDistanceRange | Qgis::ProfileGeneratorFlag::RespectsMaximumErrorMapUnit;
+}
+
+QgsPointCloudLayerProfileGenerator::~QgsPointCloudLayerProfileGenerator() = default;
+
+bool QgsPointCloudLayerProfileGenerator::generateProfile( const QgsProfileGenerationContext &context )
+{
+  if ( !mLayer || !mProfileCurve || mFeedback->isCanceled() )
+    return false;
+
+  // this is not AT ALL thread safe, but it's what QgsPointCloudLayerRenderer does !
+  // TODO: fix when QgsPointCloudLayerRenderer is made thread safe to use same approach
```

Comment:
Good point, I haven't realized the 2D renderer had this issue... Fortunately it's really just the QgsPointCloudIndex that needs to be shared among multiple threads, and the important methods are all using mutexes... but yeah, things could still break (e.g. if layer gets deleted)